allennlp/modules/vision/region_detector.py

```diff
@@ -1,31 +1,46 @@
-from typing import Dict, List, Tuple, Union
+from collections import OrderedDict
+from typing import NamedTuple, Optional, List, Tuple
 
 import torch
+from torch import nn, FloatTensor, IntTensor, Tensor
 import torch.nn.functional as F
-from torch import nn, FloatTensor, IntTensor
+import torchvision
+import torchvision.ops.boxes as box_ops
 
 from allennlp.common import Registrable
 
 
+class RegionDetectorOutput(NamedTuple):
+    features: List[Tensor]
+    boxes: List[Tensor]
+    class_probs: Optional[List[Tensor]] = None
+    class_labels: Optional[List[Tensor]] = None
+
+
 class RegionDetector(nn.Module, Registrable):
     """
-    A `RegionDetector` takes a batch of images as a tensor with the dimensions
-    (Batch, Color, Height, Width), and finds regions of interest (or "boxes") within those images.
+    A `RegionDetector` takes a batch of images, their sizes, and an ordered dictionary
+    of image features as input, and finds regions of interest (or "boxes") within those images.
 
     Those regions of interest are described by three values:
-    - A feature vector for each region, which is a tensor of shape (Batch, NumBoxes, FeatureDim)
-    - The coordinates of each region within the original image, with shape (Batch, NumBoxes, 4)
-    - (Optionally) Class probabilities from some object detector that was used to find the regions
-      of interest, with shape (Batch, NumBoxes, NumClasses).
-
-    Because the class probabilities are an optional return value, we return these tensors in a
-    dictionary, instead of a tuple (so you don't have a confusing check on the tuple size).  The
-    keys are "coordinates", "features", and "class_probs".
+
+    - `features` (`List[Tensor]`): A feature vector for each region, which is a tensor of shape
+      `(num_boxes, feature_dim)`.
+    - `boxes` (`List[Tensor]`): The coordinates of each region within the original image, with shape
+      `(num_boxes, 4)`.
+    - `class_probs` (`Optional[List[Tensor]]`): Class probabilities from some object
+      detector that was used to find the regions of interest, with shape `(num_predicted_classes,)`.
+    - `class_labels` (`Optional[List[Tensor]]`): The labels corresponding to `class_probs`.
+      Each tensor in this list has the same shape as the corresponding tensor in `class_probs`.
```

Comment:
How would we get labels?

Reply:
They are in the torchvision docs: https://pytorch.org/docs/stable/torchvision/models.html#object-detection-instance-segmentation-and-person-keypoint-detection

Reply:
I don't understand. Are these "labels" not gold annotated data from humans?